### PR TITLE
[TACHYON-877] Add TachyonPreconditions to throw Tachyon exceptions

### PR DESCRIPTION
**Pull request by miasantreble**
https://tachyon.atlassian.net/browse/TACHYON-877

### CODE REVIEW

File: common/src/main/java/tachyon/TachyonURI.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
 import java.net.URISyntaxException;
 import java.util.Arrays;
 
+import com.google.common.base.Preconditions;
```
Comment (apc999) [inline common/src/main/java/tachyon/TachyonURI.java]: NIT: import order needs to be reformatted

File: common/src/main/java/tachyon/exception/TachyonPreconditions.java [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+ * of checks! Precondition exceptions -- including those provided here, but also
+ * {@link IndexOutOfBoundsException}, {@link java.util.NoSuchElementException}, {@link
+ * UnsupportedOperationException} and others -- are used to signal that the
+ * <i>calling method</i> has made an error. This tells the caller that it should
+ * not have invoked the method when it did, with the arguments it did, or
+ * perhaps ever. Postcondition or other invariant failures should not throw
+ * these types of exceptions.
+ */
+
+public final class TachyonPreconditions{
```
Comment (calvinjia) [inline common/src/main/java/tachyon/exception/TachyonPreconditions.java]: The formatting in this file seems off, could you double check?

Comment (miasantreble) [inline common/src/main/java/tachyon/exception/TachyonPreconditions.java]: @calvinjia you mean the block comment or the missing space before the curly brace?

Comment (jsimsa) [inline common/src/main/java/tachyon/exception/TachyonPreconditions.java]: Probably both. The best way to format the code is to use either Eclipse or IntelliJ with the recommended formatters (see http://tachyon-project.org/documentation/Contributing-to-Tachyon.html). They are not flawless, but usually produce formatting that passes the "mvn checkstyle:checkstyle" check.

File: common/src/main/java/tachyon/exception/TachyonPreconditions.java [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+ * <a href="http://java.sun.com/j2se/1.5.0/docs/guide/language/assert.html"> {@code assert}</a> or a
+ * JUnit {@code assertTrue} call.
+ *
+ * <p>
+ * Take care not to confuse precondition checking with other similar types of checks! Precondition
+ * exceptions -- including those provided here, but also {@link IndexOutOfBoundsException},
+ * {@link java.util.NoSuchElementException}, {@link UnsupportedOperationException} and others -- are
+ * used to signal that the <i>calling method</i> has made an error. This tells the caller that it
+ * should not have invoked the method when it did, with the arguments it did, or perhaps ever.
+ * Postcondition or other invariant failures should not throw these types of exceptions.
```
Comment (calvinjia) [inline common/src/main/java/tachyon/exception/TachyonPreconditions.java]: I'm not sure if this is the correct doc, it seems like there are places were we do not follow this convention (for example in `removeBlockMeta`.

File: common/src/main/java/tachyon/exception/TachyonPreconditions.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+ */
+
+public final class TachyonPreconditions {
```
Comment (calvinjia) [inline common/src/main/java/tachyon/exception/TachyonPreconditions.java]: It seems like we don't really want `Preconditions` but just a helper to throw exceptions?

**Comment by apc999** [comment]
I think this PR depends on #1562 , so we can revisit this one after #1562 is merged

**Comment by apc999** [comment]
@miasantreble  A relevant PR https://github.com/amplab/tachyon/pull/1656 has been merged so Tachyon is using the new exceptions. Do you want to resume this PR? Thanks!

**Comment by yupeng9** [comment]
@miasantreble , do you plan to continue this PR? Thanks.

**Comment by miasantreble** [comment]
@yupeng9 I think the code have gone through some changes so I might create another PR instead. I will check tomorrow. Thanks for the reminder!

**Comment by yupeng9** [comment]
@miasantreble Thanks!

**Comment by miasantreble** [comment]
`tachyon.exception` package have gone through lots of changes, exceptions in this PR have either been removed or renamed and lots of new exceptions have been added. Closing this PR for now and will continue the work from a fresh branch @yupeng9 @apc999

**Comment by yupeng9** [comment]
Thanks @miasantreble. Looking forward to the new PR.

**Pull request closed without merging**